src/services/fireblocks.service.js

```diff
@@ -0,0 +1,369 @@
+    const addressInfo = await storage.getAddressInfo(address.base58);
+    if (addressInfo) {
+      ncCallerIndex = addressInfo.bip32AddressIndex;
+    }
+  }
+
+  // Now we sign
+  const indices = signingIndices.map(data => data.addressIndex);
+  if (ncCallerIndex) {
+    indices.push(ncCallerIndex);
+  }
+
+  // create signature data object from the response
+  const fbTxInfo = await client.sendRawTransaction(dataToSignHash, indices);
+  if (!fbTxInfo.signedMessages) {
+    throw new Error('Fireblocks did not return signatures');
+  }
+  const signedMessagesObj = {};
+  for (const item of fbTxInfo.signedMessages) {
+    if (item.content !== dataToSignHash.toString('hex')) {
+      throw new Error(`Fireblocks returned inconsistent signatures in transaction ${fbTxInfo.id}`);
+    }
+    if (item.derivationPath[4] in signedMessagesObj) {
+      // duplicate signature due to inputs from the same address
+      // Since the content is the same we can safely skip it
+      continue;
+    }
+    signedMessagesObj[item.derivationPath[4]] = {
+      signature: fireblocksSignatureToDER(item.signature),
+      content: item.content,
+      publicKey: item.publicKey,
+    };
+  }
+
+  for (const data of signingIndices) {
+    const sig = signedMessagesObj[data.addressIndex];
+    if (!sig) {
+      throw new Error(`Fireblocks did not return signature for address index ${data.addressIndex}`);
+    }
+    if (sig.publicKey !== data.pubkey.toString('hex')) {
+      throw new Error(`Fireblocks signature does not match locally generated public key in transaction ${fbTxInfo.id}`);
+    }
+    inputSignatures.push({
+      ...data,
+      signature: sig.signature,
+    });
+  }
+
+  if (ncCallerIndex) {
+    const sig = signedMessagesObj[ncCallerIndex];
+    if (!sig) {
+      throw new Error(`Fireblocks did not return signature for address index ${ncCallerIndex}`);
+    }
+    ncCallerSignature = sig.signature.fullSig;
+  }
+
+  return {
+    inputSignatures,
+    ncCallerSignature,
+  };
+}
+
+class FireblocksClient {
+  constructor(baseUrl, apiKey, apiSecret) {
+    this.baseUrl = baseUrl;
+    this.apiKey = apiKey;
+    this.secret = apiSecret;
+
+    this.client = axios.create({
+      baseURL: baseUrl,
+    });
+  }
+
+  /**
+   * Create the JWT authz token for a request.
+   *
+   * @param {string} path URI of the request (path + querystring)
+   * @param {Object|undefined} body Body of the request
+   * @returns {string}
+   */
+  signJWT(path, body) {
+    const nonce = uuid4();
+    const now = Math.floor(Date.now() / 1000);
+
+    const bodyJson = body ? JSON.stringify(body) : '';
+    const bodyHash = crypto.createHash('sha256').update(Buffer.from(bodyJson)).digest().toString('hex');
+
+    const headers = {
+      uri: path,
+      nonce,
+      iat: now,
+      exp: now + 300, // Adjusted to ensure it's within the required timeframe
```

Comment:
I don't know why but when using the older axios version the request is shifted by a couple minutes, so when using the usual 30s expiration it would come as

Reply:
Did some tests and found out the shift was due to the system clock being shifted and not Axios.
Don't know why the most recent Axios version works even with the shift but the expiration time has been fixed to 60s.